### PR TITLE
Podman System Reset --run-root --graph-root

### PR DESCRIPTION
**Pull request by cdoern**
added options to reinstantiate storage after resetting the podman system using
podman system reset. --run-root and --graph-root define the runRoot and graphRoot to initiate
a new runtime with. these values, once populated into a new libpd runtime, then are used to create a new storage.conf
file at the proper location (root or rootless default location).

These options allow for a more interactive and useful reset program and ill help eliminate many questions regarding how to configure the basic entities
of container storage.

depends on containers/storage#1096

Signed-off-by: cdoern <cdoern@redhat.com>

### CODE REVIEW

File: cmd/podman/system/reset.go [hunk 1]
```diff
@@ -43,6 +51,15 @@ func init() {
 	})
 	flags := systemResetCommand.Flags()
 	flags.BoolVarP(&forceFlag, "force", "f", false, "Do not prompt for confirmation")
+
+	flags.StringVar(&givenUser, "user", "", "User to setup the DB under")
```
Comment (rhatdan) [inline cmd/podman/system/reset.go:55]: What is the use of user?  I don't think this should be supported.

File: libpod/reset.go [hunk 2]
```diff
@@ -124,7 +124,20 @@ func (r *Runtime) Reset(ctx context.Context) error {
 	if storageConfPath, err := storage.DefaultConfigFile(rootless.IsRootless()); err == nil {
 		if _, err = os.Stat(storageConfPath); err == nil {
 			fmt.Printf("A storage.conf file exists at %s\n", storageConfPath)
-			fmt.Println("You should remove this file if you did not modify the configuration.")
+			if removeConf {
```
Comment (rhatdan) [inline libpod/reset.go:127]: This code should go into storage.conf, and you should just be editing the TOML file.  Not removing and creating one fresh.

File: cmd/podman/system/reset.go (hunk 1 above)
Comment (cdoern) [reply cmd/podman/system/reset.go:55]: User is kind of a quick filler option where if you can just specify something like `--user=charliedoern` and storage.conf will be populated with working paths if you broke it by putting some weird options.

Comment (rhatdan) [reply cmd/podman/system/reset.go:55]: Remove it, if I want to do that I can just do

sudo charliedoern podman system reset

Comment (cdoern) [reply cmd/podman/system/reset.go:55]: Ok, should we then remove storage.conf if `-f` is specified? If we don't remove the old storage.conf set up by the `--run-root` or `--graph-root` options then those will be used. Or I guess we can always write the new runtime config to storage.conf if it exists?

Comment (rhatdan) [reply cmd/podman/system/reset.go:55]: --force flag has nothing to do with this. storage.conf should only be rewritten if user specifies the --run-root and --graph-root options.

File: docs/source/markdown/podman-system-reset.1.md [hunk 3]
```diff
@@ -26,7 +26,28 @@ Do not prompt for confirmation
 
 Print usage statement
 
+#### **--run-root**
+
+Establishes a new run root directory and storage.conf file once the system is reset, cannot be used in conjunction with --user
+
+#### **--graph-root**
+
+Establishes a new graph root directory and sotrage.conf file once the system is reset, cannot be used in conjunction with --user
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-reset.1.md:35]:
```suggestion
Establishes a new graph root directory and storage.conf file once the system is reset, cannot be used in conjunction with --user
```

File: docs/source/markdown/podman-system-reset.1.md [hunk 4]
```diff
@@ -26,7 +26,28 @@ Do not prompt for confirmation
 
 Print usage statement
 
+#### **--run-root**
+
+Establishes a new run root directory and storage.conf file once the system is reset, cannot be used in conjunction with --user
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-reset.1.md:31]:
```suggestion
Establishes a new run root directory and storage.conf file once the system is reset, cannot be used in conjunction with --user.
```

File: docs/source/markdown/podman-system-reset.1.md [hunk 5]
```diff
@@ -26,7 +26,28 @@ Do not prompt for confirmation
...
+
+#### **--graph-root**
+
+Establishes a new graph root directory and sotrage.conf file once the system is reset, cannot be used in conjunction with --user
+
+#### **--user**
+
+Establishes new run and graph root directories and storage.conf file once the system is reset. the run and graph root directories are generated by retrieving rhe user's home and run directory.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-reset.1.md:39]:
```suggestion
Establishes new run and graph root directories and storage.conf file once the system is reset. The run and graph root directories are generated by retrieving the user's home and run directory.
```

File: docs/source/markdown/podman-system-reset.1.md [hunk 6]
```diff
@@ -26,7 +26,28 @@ Do not prompt for confirmation
...
+
 ## EXAMPLES
+Reset the system and create a new storage.conf enerated from the given user's defaults
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-reset.1.md:42]:
```suggestion
Reset the system and create a new storage.conf file generated from the given user's defaults
```

File: docs/source/markdown/podman-system-reset.1.md [hunk 7]
```diff
@@ -26,7 +26,28 @@ Do not prompt for confirmation
...
+```
+$ podman system reset --user=1000:1000
+```
+
+Reset the system and create a new storage.conf with the given run and graph root
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-reset.1.md:47]:
```suggestion
Reset the system and create a new storage.conf file with the given run and graph root.
```

File: test/system/800-system-reset.bats [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    # In case the active-volumes test failed: clean up stray volumes
+    run_podman volume rm -a
+}
+
+
+@test "podman system reset - storage options" {
+    whoami=$(id -un)
+    if ! is_rootless; then
+        run_podman system reset -f --user=$whoami
+        if [ "$status" -ne 0 ]; then
```
Comment (edsantiago) [inline test/system/800-system-reset.bats:29]: `run_podman` does implicit exit-status checks: you do not (and should not) need to check `$status`. (With a very narrow exception that does not apply here).

What you _should_ always do is check `$output`.

Comment (cdoern) [reply test/system/800-system-reset.bats:29]: ok thanks, that makes more sense.Two questions on these tests: 1) is it ok to change the run root and graph root here or will it change the entire system's? and 2) I was unsure of what to name the file is `800-` an ok number to use?

Comment (edsantiago) [reply test/system/800-system-reset.bats:29]: Changing `--root` and `--runroot` (no hyphen) is always (AFAIK) safe. There are a handful of cases where we already do that in the test. `graphroot`, I have no idea: this is the first I've ever heard of that option, and I don't even see it in `--help`.

`800` is a great choice for something as scary as `system reset`.

Comment (edsantiago) [reply test/system/800-system-reset.bats:29]: Of course. `grep 'is.*\.\*' test/system/*bats`

**Comment by cdoern** [comment]
brought in a bunch of stuff when getting the new c/storage release, not sure if I grabbed anything extra or not...

File: test/system/800-system-reset.bats [hunk 9]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bats   -*- bats -*-
+#
+# tests for podman system reset
+#
+
+load helpers
+
+function setup() {
+    if is_remote; then
+        skip "only applicable when running non remote"
```
Comment (rhatdan) [inline test/system/800-system-reset.bats:10]: Nit use
skip_if_remote  "only applicable when running non remote"

File: test/system/800-system-reset.bats [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+@test "podman system reset - storage options" {
+        rr=$PODMAN_TMPDIR/containers
+        mkdir -p $rr
+
+        gr=$PODMAN_TMPDIR/.local/share/containers/storage
+        mkdir -p $gr
+
+        run_podman system reset -f --run-root=${rr} --graph-root=${gr}
+        dir=/home/$(id -un)/.config/containers/storage.conf
```
Comment (rhatdan) [inline test/system/800-system-reset.bats:30]: I don't believe Ed is going to like this. You might be able to use a STORAGE_CONF environment variable for this
Or at least set $HOME=PODMAN_TMPDIR/home or something like that.

Final Nit is storage.conf is a file not a dir.  Would more sense to say storage_file=/home...

File: pkg/domain/infra/runtime_libpod.go [hunk 11]
```diff
@@ -126,12 +127,17 @@ func getRuntime(ctx context.Context, fs *flag.FlagSet, opts *engineOpts) (*libpo
 		storageSet = true
 	}
 
-	if fs.Changed("root") {
+	gr, _ := fs.GetString("graph-root")
+	if fs.Changed("root") || len(gr) > 0 {
```
Comment (rhatdan) [inline pkg/domain/infra/runtime_libpod.go:131]: I don't see the value of gr being used?

File: pkg/domain/infra/runtime_libpod.go [hunk 12]
```diff
@@ -126,12 +127,17 @@ func getRuntime(ctx context.Context, fs *flag.FlagSet, opts *engineOpts) (*libpo
...
+		options = append(options, libpod.WithNewDB())
 		storageSet = true
 		storageOpts.GraphRoot = cfg.Engine.StaticDir
 		storageOpts.GraphDriverOptions = []string{}
 	}
-	if fs.Changed("runroot") {
+
+	rr, _ := fs.GetString("run-root")
+	if fs.Changed("runroot") || len(rr) > 0 {
```
Comment (rhatdan) [inline pkg/domain/infra/runtime_libpod.go:139]: Where else is rr being used?

File: libpod/reset.go [hunk 13]
```diff
@@ -122,9 +123,23 @@ func (r *Runtime) Reset(ctx context.Context) error {
 		}
 	}
 	if storageConfPath, err := storage.DefaultConfigFile(rootless.IsRootless()); err == nil {
-		if _, err = os.Stat(storageConfPath); err == nil {
+		if _, err := os.Stat(storageConfPath); err == nil {
 			fmt.Printf("A storage.conf file exists at %s\n", storageConfPath)
-			fmt.Println("You should remove this file if you did not modify the configuration.")
+			fmt.Println("you can remove it or override it with --run-root and --graph-root")
```
Comment (rhatdan) [inline libpod/reset.go:128]: I don't like this change, I want users to remove the storage.conf file, and not use --run-root or graphroot.
Also if I am doing updateConf, why print these messages at all.

File: test/system/800-system-reset.bats (hunk 10 above)
Comment (edsantiago) [reply test/system/800-system-reset.bats:30]: Thanks for the ping. Dan is correct, I would really not appreciate having my podman environment clobbered by running a test. Also, `/home/$(id -un)` doesn't work for root. Also, I too am confused by your use of `$dir` and the `[ -d` below. I need to take a pretty close look at this, and will try to do so today.

File: libpod/reset.go (hunk 13 above)
Comment (cdoern) [reply libpod/reset.go:128]: Ok, I can revert this message

File: pkg/domain/infra/runtime_libpod.go (hunk 12 above)
Comment (cdoern) [reply pkg/domain/infra/runtime_libpod.go:139]: rr is just being used below in `if fs.Changed("runroot") || len(rr) > 0 {` the issue is I cannot just do `fs.Changed` since the flag will not always exist at runtime.

File: pkg/domain/infra/runtime_libpod.go (hunk 11 above)
Comment (cdoern) [reply pkg/domain/infra/runtime_libpod.go:131]: rgr is just being used below in `if fs.Changed("root") || len(gr) > 0 {` the issue is I cannot just do `fs.Changed` since the flag will not always exist at runtime.

File: test/system/800-system-reset.bats (hunk 10 above)
Comment (cdoern) [reply test/system/800-system-reset.bats:30]: I was unsure of how to approach this, is it possible this same test could work as integration? I put it in system due to the nature of modifying storage.conf. This creates a similar issue as scp does with modifying user data.

File: pkg/domain/infra/runtime_libpod.go (hunk 11 above)
Comment (rhatdan) [reply pkg/domain/infra/runtime_libpod.go:131]: Huh?  fs.Changed("root") the --root not existing?

Comment (cdoern) [reply pkg/domain/infra/runtime_libpod.go:131]: no `--graph-root` is a non-global system reset flag but many different commands go through `getRuntime`. If i always grab the flag via `fs.Changed` and it does not exist in the command being run, we error out

**Comment by rhatdan** [comment]
@cdoern  Is this something we need to get into podman 4.0, or can it wait for podman 4.1?

**Comment by cdoern** [comment]
@rhatdan I think integrations tests might work here. Is there a way to set my storage.conf via an env variable or a global podman option? If so I can make a dummy file and write/read from that within the testing suite

**Comment by rhatdan** [comment]
CONTAINERS_STORAGE_CONF=/PATH/storage.conf podman ...

man podman
...
   CONTAINERS_STORAGE_CONF
       Set default location of the storage.conf file.

**Comment by cdoern** [comment]
@rhatdan is there any reason `types.Save` in `libpod/reset.go` removes and does not update the storage.conf? I am testing this locally now and it seems to just remove the file but the function looks right to me.

**Comment by rhatdan** [comment]
It should work, I have a feeling, you are not saving what you think you are.

**Comment by cdoern** [comment]
@rhatdan can we cut a new c/storage release so I can fix this up?

**Comment by rhatdan** [comment]
You do not need a release.  Just take the current checksum and use that rather then a version.

```
$ git diff go.mod
diff --git a/go.mod b/go.mod
index 79880469b..f6a2e0f1d 100644
--- a/go.mod
+++ b/go.mod
@@ -17,7 +17,7 @@ require (
        github.com/containers/image/v5 v5.19.1
        github.com/containers/ocicrypt v1.1.2
        github.com/containers/psgo v1.7.2
-       github.com/containers/storage v1.38.2
+       github.com/containers/storage 77cf15bc7897a70cd2e0b026ac1fe9757b791075
        github.com/coreos/go-systemd/v22 v22.3.2
        github.com/coreos/stream-metadata-go v0.0.0-20210225230131-70edb9eb47b3
        github.com/cyphar/filepath-securejoin v0.2.3
```

File: libpod/boltdb_state.go [hunk 14]
```diff
@@ -408,12 +408,29 @@ func (s *BoltState) ValidateDBConfig(runtime *Runtime) error {
 	}
 	defer s.deferredCloseDBCon(db)
 
+	//runRootName := "storage temporary directory (runroot)"
+	//runRootKey := []byte(runRootName)
 	// Check runtime configuration
 	if err := checkRuntimeConfig(db, runtime); err != nil {
+		/*	if runtime.newDB {
```
Comment (rhatdan) [inline libpod/boltdb_state.go:415]: Why is this commented code here?  If it is not going to be used, remove it.  Git will remember the code.

File: libpod/reset.go [hunk 15]
```diff
@@ -139,9 +140,25 @@ func (r *Runtime) Reset(ctx context.Context) error {
...
```
Comment (rhatdan) [inline libpod/reset.go:145]: you can remove it or override it with the --run-root and --graph-root options.

**Review by rhatdan — APPROVED** [review]

File: pkg/domain/infra/runtime_libpod.go (hunk 11 above)
Comment (jwhonce) [reply pkg/domain/infra/runtime_libpod.go:131]: Using `fs.Lookup("graph-root")` might be clearer. It returns nil if the flag is not in the FlagSet.

**Comment by cdoern** [comment]
@containers/podman-maintainers PTAL

File: docs/source/markdown/podman-system-reset.1.md [hunk 16]
```diff
@@ -7,7 +7,7 @@ podman\-system\-reset - Reset storage back to initial state
 **podman system reset** [*options*]
 
 ## DESCRIPTION
-**podman system reset** removes all pods, containers, images, networks and volumes.
+**podman system reset** removes all pods, containers, images and volumes. can also be used to instantiate new directories for storing images and information about Podmans's state.
```
Comment (Luap99) [inline docs/source/markdown/podman-system-reset.1.md:10]:
```suggestion
**podman system reset** removes all pods, containers, images and volumes. Can also be used to instantiate new directories for storing images and information about Podmans's state.
```

Comment (Luap99) [inline docs/source/markdown/podman-system-reset.1.md:10]:
```suggestion
**podman system reset** removes all pods, containers, images and volumes. Can also be used to instantiate new directories for storing images and information about Podmans's state.
```

File: libpod/boltdb_state.go [hunk 17]
```diff
@@ -413,7 +413,7 @@ func (s *BoltState) ValidateDBConfig(runtime *Runtime) error {
 		return err
 	}
 
-	return nil
+	return err
```
Comment (Luap99) [inline libpod/boltdb_state.go:416]: Why is this changed, if anything this should just be `return checkRuntimeConfig(db, runtime)`

Comment (Luap99) [inline libpod/boltdb_state.go:416]: Why is this changed, if anything this should just be `return checkRuntimeConfig(db, runtime)`

File: libpod/boltdb_state_internal.go [hunk 18]
```diff
@@ -157,10 +157,10 @@ func checkRuntimeConfig(db *bolt.DB, rt *Runtime) error {
 
 		for _, check := range checks {
 			exists, err := readOnlyValidateConfig(configBkt, check)
-			if err != nil {
+			if err != nil && (err != define.ErrDBBadConfig && !rt.newDB) {
```
Comment (Luap99) [inline libpod/boltdb_state_internal.go:160]: please add a comment why this is necessary

Comment (Luap99) [inline libpod/boltdb_state_internal.go:160]: please add a comment why this is necessary

**Review by Luap99 — CHANGES_REQUESTED** [review]
I have tested this change and it does not work!

```
$ mkdir -p $XDG_RUNTIME_DIR/podman-test ~/podman-test
$ bin/podman system reset --run-root $XDG_RUNTIME_DIR/podman-test --graph-root ~/podman-test
WARNING! This will remove:
        - all containers
        - all pods
        - all images
        - all networks
        - all build cache
Are you sure you want to continue? [y/N] y
$ ls ~/podman-test
ls: cannot access '/home/pholzing/podman-test': No such file or directory
$ ls $XDG_RUNTIME_DIR/podman-test
ls: cannot access '/run/user/1000/podman-test': No such file or directory
$ bin/podman info --format {{.Store.GraphRoot}}
/home/pholzing/.local/share/containers/storage
```

**Comment by cdoern** [comment]
@Luap99 it looks like yours did not work because you do not have a storage.conf. We only honor the flags if the file exists int the default dir. I can add a warning for that

**Comment by rhatdan** [comment]
Should we support his use case?

File: cmd/podman/system/reset.go [hunk 19]
```diff
@@ -44,6 +46,12 @@ func init() {
 	})
 	flags := systemResetCommand.Flags()
 	flags.BoolVarP(&forceFlag, "force", "f", false, "Do not prompt for confirmation")
+
+	flags.StringVar(&runDir, "run-root", "", "run root directory to use for storage")
```
[the remark this answers is not in the document]
Comment (cdoern) [reply cmd/podman/system/reset.go:50]: Does this make sense to do?

That could get confusing with;

https://github.com/containers/podman/blob/4967dee87f198fea000c0094a3f703c0124c7830/pkg/domain/infra/runtime_libpod.go#L130-L141

right? also are you allowed to have two flags with the same name?

**Comment by cdoern** [comment]
@Luap99 I addressed many of your review comments with some questions on some. I guess it is up to you really if you want the use case supported.

I think the reason we opted not to @rhatdan is because then people could pass these options unknowingly and mess up their storage config in a pretty major way when there was none to begin with.

File: cmd/podman/system/reset.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (vrothberg) [reply cmd/podman/system/reset.go:50]: I concur with @Luap99.  AFAIK, duplicates are allowed.

**Review by vrothberg — COMMENTED** [review]
Nice work, @cdoern !

**Comment by Luap99** [comment]
IMO the logic should do a proper reset and only then create the new the graph/runroot. If you do this you can avoid all the db conflicts that your are currently handling in your code.

**Comment by cdoern** [comment]
@Luap99 so your recommendations are 1) fix the toml writing in c/storage and 2) make this write even when a config file does not exist?

**Comment by Luap99** [comment]
Well first I am not sure if this feature is really needed since you could just add the config value manually as user.
But I guess we have some form of request for this otherwise you would not work on this.

1) yes you need to fix this in c/storage
2) I would think that this makes more sense

But more importantly this does not work yet because you are not actually deleting to old location. IMO this must do a full reset and only then create the new storage.conf. I do not see why we need all this tricky special db handling here.

**Comment by cdoern** [comment]
@Luap99 I worked on this because I thought reset was lacking customization. If we think this only introduces more issues I guess I can close this. I was just trying to create more options to make the feature more helpful.

@rhatdan WDYT? Should we scrap this or is there use in it?

**Comment by rhatdan** [comment]
Lets scrap it. Because user can just reset, and then edit storage.conf to do the changes you want correct?

**Comment by cdoern** [comment]
Yeah you can just do the changes by hand, I will close this
